soccer/planning/InterpolatedPath.cpp

```diff
@@ -190,69 +195,76 @@ void InterpolatedPath::draw(DebugDrawer* const debug_drawer,
 
     for (const Entry& entry : waypoints) {
         Packet::DebugRobotPath::DebugRobotPathPoint* pt = dbg->add_points();
-        *pt->mutable_pos() = entry.pos();
-        *pt->mutable_vel() = entry.vel();
+        *pt->mutable_pos() = entry.pose.position();
+        *pt->mutable_vel() = entry.vel.linear();
     }
 }
 
 std::optional<RobotInstant> InterpolatedPath::eval(RJ::Seconds t) const {
-    if (t < RJ::Seconds::zero()) {
-        return RobotInstant(waypoints.front().instant);
-    }
-    //    if (t < RJ::Seconds::zero()) {
-    //        debugThrow(
-    //            invalid_argument("A time less than 0 was entered for time t.
-    //            t=" + to_string(t)));
-    //    }
-    /*
-    float linearPos;
-    float linearSpeed;
-    bool pathIsValid = TrapezoidalMotion(
-        length(),
-        maxSpeed,
-        maxAcceleration,
-        t,
-        startSpeed,
-        endSpeed,
-        linearPos,      //  these are set by reference since C++ can't return
-    multiple values
-        linearSpeed);   //
-
-    Point direction;
-    if(!getPoint(linearPos, targetPosOut, direction)) {
-        return false;
-    }
-
-    targetVelOut = direction * linearSpeed;
-    */
-    if (waypoints.size() == 0 || waypoints.size() == 1) {
+    if (waypoints.size() < 2) {
         return std::nullopt;
     }
-    if (t < waypoints[0].time) {
-        debugThrow(
-            invalid_argument("The start time should not be less than zero"));
+    if (t <= waypoints.front().time) {
+        return waypoints.front().instant();
+    } else if (t > waypoints.back().time) {
+        return std::nullopt;
+    } else if (t == waypoints.back().time) {
+        return waypoints.back().instant();
     }
 
-    int i = 0;
-    while (waypoints[i].time <= t) {
-        if (waypoints[i].time == t) {
-            return RobotInstant(waypoints[i].instant);
-        }
-        i++;
-        if (i == size()) {
-            return std::nullopt;
+    std::vector<Entry>::const_iterator prev_it = waypoints.begin();
+    std::vector<Entry>::const_iterator next_it = waypoints.begin();
+    while (next_it != waypoints.end()) {
+        prev_it = next_it;
+        next_it++;
+
+        if (next_it->time >= t) {
+            break;
         }
     }
-    RJ::Seconds deltaT = (waypoints[i].time - waypoints[i - 1].time);
-    if (deltaT == RJ::Seconds::zero()) {
-        return RobotInstant(waypoints[i].instant);
-    }
-    float constant = (t - waypoints[i - 1].time) / deltaT;
 
-    return RobotInstant(MotionInstant(waypoints[i - 1].pos() * (1 - constant) +
-                                          waypoints[i].pos() * (constant),
-                                      waypoints[i - 1].vel() * (1 - constant) +
-                                          waypoints[i].vel() * (constant)));
+    assert(prev_it != waypoints.end());
+    assert(next_it != waypoints.end());
```

Comment:
Due to the if statements above, this should never happen, but if for some reason we get to this while loop when `t > waypoints.back().time`, it will possibly crash on `next_it->time` inside the if statement on line 221 instead of this assert. If `next_it == waypoints.back()`, it enters the loop, increments `next_it` to `waypoints.end()` then tries to dereference a past-the-end iterator resulting in undefined behavior.

https://repl.it/repls/BigheartedAthleticConversion

Reply:
Good catch. I think this can be fixed by moving the check above the iterator increments.